nav2_loopback_sim/nav2_loopback_sim/loopback_simulator.py

```diff
@@ -0,0 +1,222 @@
+from sensor_msgs.msg import LaserScan
+from tf2_ros import TransformBroadcaster
+import tf_transformations
+
+from .utils import addYawToQuat, matrixToTransform, transformStampedToMatrix
+
+
+"""
+This is a loopback simulator that replaces a physics simulator to create a
+frictionless, inertialess, and collisionless simulation environment. It
+accepts cmd_vel messages and publishes odometry & TF messages based on the
+cumulative velocities received to mimick global localization and simulation.
+It also accepts initialpose messages to set the initial pose of the robot
+to place anywhere.
+"""
+
+
+class LoopbackSimulator(Node):
+
+    def __init__(self):
+        super().__init__(node_name='loopback_simulator')
+        self.curr_cmd_vel = None
+        self.curr_cmd_vel_time = self.get_clock().now()
+        self.initial_pose = None
+        self.timer = None
+        self.setupTimer = None
+
+        self.declare_parameter('update_duration', 0.01)
+        self.update_dur = self.get_parameter('update_duration').get_parameter_value().double_value
+
+        self.declare_parameter('base_frame_id', 'base_footprint')
+        self.base_frame_id = self.get_parameter('base_frame_id').get_parameter_value().string_value
+
+        self.declare_parameter('map_frame_id', 'map')
+        self.map_frame_id = self.get_parameter('map_frame_id').get_parameter_value().string_value
+
+        self.declare_parameter('odom_frame_id', 'odom')
+        self.odom_frame_id = self.get_parameter('odom_frame_id').get_parameter_value().string_value
+
+        self.declare_parameter('scan_frame_id', 'base_scan')
+        self.scan_frame_id = self.get_parameter('scan_frame_id').get_parameter_value().string_value
+
+        self.t_map_to_odom = TransformStamped()
+        self.t_map_to_odom.header.frame_id = self.map_frame_id
+        self.t_map_to_odom.child_frame_id = self.odom_frame_id
+        self.t_odom_to_base_link = TransformStamped()
+        self.t_odom_to_base_link.header.frame_id = self.odom_frame_id
+        self.t_odom_to_base_link.child_frame_id = self.base_frame_id
+
+        self.tf_broadcaster = TransformBroadcaster(self)
+
+        self.initial_pose_sub = self.create_subscription(
+            PoseWithCovarianceStamped,
+            'initialpose', self.initialPoseCallback, 10)
+        self.cmd_vel_sub = self.create_subscription(
+            Twist,
+            'cmd_vel', self.cmdVelCallback, 10)
+        self.odom_pub = self.create_publisher(Odometry, 'odom', 10)
+
+        sensor_qos = QoSProfile(
+            reliability=ReliabilityPolicy.BEST_EFFORT,
+            durability=DurabilityPolicy.VOLATILE,
+            depth=10)
+        self.scan_pub = self.create_publisher(LaserScan, 'scan', sensor_qos)
+
+        self.setupTimer = self.create_timer(0.1, self.setupTimerCallback)
+        self.info('Loopback simulator initialized')
+
+    def setupTimerCallback(self):
+        # Publish initial identity odom transform & laser scan to warm up system
```

Comment:
I mean for it to be repeated on a timer that runs before the initial pose is set. The 'real' simulators or hardware robots will have their odometry systems continuously publishing and that `odom->base_link` is needed by many node's lifecycle startup processes to make Nav2 fully initialize. We could publish it once, but that would be fragile in case there's a delay or a long plugin initialization cycle for that message to be out of the TF buffer duration. Plus, this is how "real" robots and simulators operate so its analogoous 

The comment is mostly there for tutorial value for later on-lookers. I agree most of my comments are not necessary